pallets/proof-of-stake/src/lib.rs

```diff
@@ -0,0 +1,916 @@
+			<T as Config>::PoolPromoteApi::get_pool_rewards_v2(liquidity_asset_id)
+				.ok_or_else(|| DispatchError::from(Error::<T>::NotAPromotedPool))?;
+
+		let rewards_info: RewardInfo = Self::get_rewards_info(user.clone(), liquidity_asset_id);
+
+		let last_checkpoint = rewards_info.last_checkpoint;
+		let pool_ratio_at_last_checkpoint = rewards_info.pool_ratio_at_last_checkpoint;
+		let missing_at_last_checkpoint = rewards_info.missing_at_last_checkpoint;
+		let liquidity_assets_amount: Balance = rewards_info.activated_amount;
+
+		let time_passed = current_time
+			.checked_sub(last_checkpoint)
+			.ok_or_else(|| DispatchError::from(Error::<T>::PastTimeCalculation))?;
+
+		if time_passed == 0 {
+			pool_ratio_current = pool_ratio_at_last_checkpoint;
+		}
+
+		let missing_at_checkpoint_new =
+		Self::calculate_missing_at_checkpoint_v2(time_passed, missing_at_last_checkpoint)?;
+
+		let user_current_rewards = Self::calculate_rewards_v2(
+			liquidity_assets_amount,
+			liquidity_asset_id,
+			last_checkpoint,
+			pool_ratio_at_last_checkpoint,
+			missing_at_last_checkpoint,
+			pool_ratio_current,
+		)?;
+
+		let activated_amount_new = liquidity_assets_amount
+			.checked_sub(liquidity_assets_burned)
+			.ok_or_else(|| DispatchError::from(Error::<T>::LiquidityCheckpointMathError))?;
+
+		let activated_amount_new_u256: U256 = activated_amount_new.into();
+
+		let missing_at_checkpoint_after_burn: U256 = activated_amount_new_u256
+			.checked_mul(missing_at_checkpoint_new)
+			.ok_or_else(|| DispatchError::from(Error::<T>::LiquidityCheckpointMathError))?
+			.checked_div(liquidity_assets_amount.into())
+			.ok_or_else(|| DispatchError::from(Error::<T>::DivisionByZero))?;
+
+		let total_available_rewards = user_current_rewards
+			.checked_add(rewards_info.rewards_not_yet_claimed)
+			.ok_or_else(|| DispatchError::from(Error::<T>::LiquidityCheckpointMathError))?
+			.checked_sub(rewards_info.rewards_already_claimed)
+			.ok_or_else(|| DispatchError::from(Error::<T>::LiquidityCheckpointMathError))?;
+
+		let rewards_info_new: RewardInfo = RewardInfo {
+			activated_amount: activated_amount_new,
+			rewards_not_yet_claimed: total_available_rewards,
+			rewards_already_claimed: 0_u128,
+			last_checkpoint: current_time,
+			pool_ratio_at_last_checkpoint: pool_ratio_current,
+			missing_at_last_checkpoint: missing_at_checkpoint_after_burn,
+		};
+
+		RewardsInfo::<T>::insert(user.clone(), liquidity_asset_id, rewards_info_new);
+
+		LiquidityMiningActivePoolV2::<T>::try_mutate(liquidity_asset_id, |active_amount| {
+			if let Some(val) = active_amount.checked_sub(liquidity_assets_burned) {
+				*active_amount = val;
+				Ok(())
+			} else {
+				Err(())
+			}
+		})
+		.map_err(|_| DispatchError::from(Error::<T>::LiquidityCheckpointMathError))?;
+
+		<T as Config>::ActivationReservesProvider::deactivate(
+			liquidity_asset_id,
+			&user,
+			liquidity_assets_burned,
+		);
+
+		Ok(())
+	}
+}
+
+impl<T: Config> XykRewardsApi for Pallet<T> {
```

Comment:
i would rather get rid of `XYK` from the trait name, maybe sth like CumulativeWorkRewardsApi ?

Reply:
Done